src/dbnode/storage/namespace.go

```diff
@@ -1196,21 +1196,33 @@ func (n *dbNamespace) ColdFlush(flushPersist persist.FlushPreparer) error {
 		return err
 	}
 
+	// NB(bodu): Deferred shard cold flushes so that we can ensure that cold flush index data is
+	// persisted before persisting TSDB data to ensure crash consistency.
+	shardColdFlushes := make([]ShardColdFlush, 0, len(shards))
 	for _, shard := range shards {
-		err := shard.ColdFlush(flushPersist, resources, nsCtx, onColdFlushNs)
+		shardColdFlush, err := shard.ColdFlush(flushPersist, resources, nsCtx, onColdFlushNs)
 		if err != nil {
 			detailedErr := fmt.Errorf("shard %d failed to compact: %v", shard.ID(), err)
 			multiErr = multiErr.Add(detailedErr)
-			// Continue with remaining shards.
+			continue
 		}
+		shardColdFlushes = append(shardColdFlushes, shardColdFlush)
 	}
 
-	if err := onColdFlushNs.Done(); err != nil {
-		multiErr = multiErr.Add(err)
+	if multiErr.NumErrors() == 0 {
+		if err := onColdFlushNs.Done(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
 	}
-	if onColdFlushDone != nil {
+	if multiErr.NumErrors() == 0 && onColdFlushDone != nil {
 		multiErr = multiErr.Add(onColdFlushDone())
 	}
+	for _, shardColdFlush := range shardColdFlushes {
+		if multiErr.NumErrors() != 0 {
```

Comment:
nit; you can do this check outside of the loop

Reply:
Yeah true, I guess we probably want to try to call "all dones" if we can.

Reply:
Ah good point; then I think there's a bug here, since it'll continue as soon as the first one fails

Reply:
Then what should the expected behaviour be? We can't roll back a cold flush for already flushed shards and the logic previously was basically the same. Attempt to flush each shard and log errors for failed shards.

Reply:
I think expected behaviour is to call done on each flush, and just gather all the errors?
```golang
for _, shardColdFlush := range shardColdFlushes {
 multiErr = multiErr.Add(shardColdFlush.Done())
}
```

Reply:
Hmm, I'll move the error check around. I think we need to check the error still before attempting to complete shard cold flushes. Since if the index cold flusher fails, we don't want to complete the shard cold flush since those series are not persisted in the index yet.